flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteStateTableSnapshot.java

```diff
@@ -217,12 +223,29 @@ protected void reportAllElementKeyGroups() {
 			int flattenIndex = 0;
 			for (CopyOnWriteStateTable.StateTableEntry<K, N, S> entry : partitioningDestination) {
 				while (null != entry) {
-					final int keyGroup = KeyGroupRangeAssignment.assignToKeyGroup(entry.key, totalKeyGroups);
-					reportKeyGroupOfElementAtIndex(flattenIndex, keyGroup);
-					partitioningSource[flattenIndex++] = entry;
+					CopyOnWriteStateTable.StateTableEntry<K, N, S> filteredEntry = filterEntry(entry);
```

Comment:
What I don't like about this line in particular is that it is a bit invasive and adds branching to an otherwise very tight loop, and I am not sure that this will simply be removed by the JIT. Can we do this in a way that does not involve filtering-calls if it is clear that no filtering is done, similar to the `NestedMapsStateTable`?

Overall, maybe we can keep in mind that there will be a unified format for the different state backends, and that probably will not start by writing a count. After that is introduced, we can also push the filtering right before the writing to the stream for a nice separation.

Reply:
I think we can extend this class with filtering and use extension only if there is an actual filter.